src/actions/DepositERC4626Action.sol

```diff
@@ -0,0 +1,118 @@
+
+/// @title DepositERC4626Action
+/// @author Otim Labs, Inc.
+/// @notice an Action that deposits ERC20 tokens into an ERC4626 vault
+contract DepositERC4626Action is IAction, IDepositERC4626Action, Interval, OtimFee {
+    constructor(address feeTokenRegistryAddress, address treasuryAddress, uint256 gasConstant_)
+        OtimFee(feeTokenRegistryAddress, treasuryAddress, gasConstant_)
+    {}
+
+    /// @inheritdoc IAction
+    function argumentsHash(bytes calldata arguments) public pure returns (bytes32, bytes32) {
+        return (INSTRUCTION_TYPEHASH, hash(abi.decode(arguments, (DepositERC4626))));
+    }
+
+    /// @inheritdoc IDepositERC4626Action
+    function hash(DepositERC4626 memory arguments) public pure returns (bytes32) {
+        return keccak256(
+            abi.encode(
+                ARGUMENTS_TYPEHASH,
+                arguments.vault,
+                arguments.recipient,
+                arguments.value,
+                arguments.minTotalAssets,
+                hash(arguments.schedule),
+                hash(arguments.fee)
+            )
+        );
+    }
+
+    /// @inheritdoc IAction
+    function execute(
+        InstructionLib.Instruction calldata instruction,
+        InstructionLib.Signature calldata,
+        InstructionLib.ExecutionState calldata executionState
+    ) external override returns (bool) {
+        // initial gas measurement for fee calculation
+        uint256 startGas = gasleft();
+
+        // decode the arguments from the instruction
+        DepositERC4626 memory arguments = abi.decode(instruction.arguments, (DepositERC4626));
+
+        // if first execution, validate the input
+        if (executionState.executionCount == 0) {
+            if (
+                arguments.recipient == address(0) || arguments.vault == address(0) || arguments.value == 0
+                    || arguments.minTotalAssets == 0
+            ) {
+                revert InvalidArguments();
+            }
+
+            checkStart(arguments.schedule);
+        } else {
+            checkInterval(arguments.schedule, executionState.lastExecuted);
+        }
+
+        // get the underlying token
+        address underlyingToken = IERC4626(arguments.vault).asset();
+
+        // get the max deposit amount
+        uint256 maxDeposit = IERC4626(arguments.vault).maxDeposit(arguments.recipient);
+
+        // if the max deposit amount is zero, revert
+        if (maxDeposit == 0) {
+            revert MaxDepositZero();
+        }
+
+        // initialize deposit amount
+        uint256 depositAmount = arguments.value;
+
+        // if the max deposit amount is less than the deposit amount,
+        // set the deposit amount to the max deposit amount and emit an event
+        if (depositAmount > maxDeposit) {
+            depositAmount = maxDeposit;
+
+            emit MaxDepositReached(maxDeposit);
+        }
+
+        // check if the account has enough balance to deposit
```

Comment:
nit: worth including this check earlier to fast-fail immediately prior to making any adjustments if the source balance is insufficient and the call basically doomed to fail?

Reply:
Also since the system depends on simulating before executing, we only really need to optimize for the happy path (total gas used) and not partial paths since we don't spend any gas on Instructions that revert